src/Agent.Worker/AgentPluginManager.cs

```diff
@@ -0,0 +1,332 @@
+using Microsoft.TeamFoundation.DistributedTask.WebApi;
+using Agent.PluginCore;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Microsoft.VisualStudio.Services.WebApi;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.Loader;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker
+{
+    [ServiceLocator(Default = typeof(AgentPluginManager))]
+    public interface IAgentPluginManager : IAgentService
+    {
+        Dictionary<Guid, Dictionary<string, Definition>> SupportedTasks { get; }
+        Dictionary<string, Dictionary<string, AgentCommandPluginInfo>> SupportedLoggingCommands { get; }
+
+        Task RunPluginTaskAsync(IExecutionContext context, string plugin, Dictionary<string, string> inputs, Dictionary<string, string> environment, string stage, EventHandler<ProcessDataReceivedEventArgs> outputHandler);
+        void ProcessCommand(IExecutionContext context, Command command);
+    }
+
+    public sealed class AgentPluginManager : AgentService, IAgentPluginManager
+    {
+        private readonly Dictionary<string, Dictionary<string, AgentCommandPluginInfo>> _supportedLoggingCommands = new Dictionary<string, Dictionary<string, AgentCommandPluginInfo>>(StringComparer.OrdinalIgnoreCase);
+        private readonly Dictionary<Guid, Dictionary<string, Definition>> _supportedTasks = new Dictionary<Guid, Dictionary<string, Definition>>();
+
+        private readonly List<AgentPluginInfo> _taskPlugins = new List<AgentPluginInfo>()
+        {
+            new AgentPluginInfo()
+            {
+                AgentPluginAssembly = "Agent.RepositoryPlugin",
+                AgentPluginEntryPoint = "Agent.RepositoryPlugin.CheckoutTask"
+            }
+        };
+
+        private readonly List<AgentPluginInfo> _commandPlugins = new List<AgentPluginInfo>()
+        {
+            new AgentPluginInfo()
+            {
+                AgentPluginAssembly = "Agent.DropPlugin",
+                AgentPluginEntryPoint = "Agent.DropPlugin.ArtifactUploadCommand"
+            }
+        };
+
+        public Dictionary<string, Dictionary<string, AgentCommandPluginInfo>> SupportedLoggingCommands => _supportedLoggingCommands;
```

Comment:
Rather than expose these dictionaries, a nicer interface might be something like GetCommand() and return null if not found

Reply:
GetTask(guid, version)
GetCommnad(String, string)

Reply:
fixed.